presentation/src/main/java/org/sagebionetworks/research/presentation/speech/TextToSpeechService.kt

```diff
@@ -0,0 +1,420 @@
+        // We observe the state so that if every client unbinds and the service stops speaking, the service
+        // stops itself. In the initial state the service is IDLE, and has 0 bound clients.
+        cleanupMediator.postValue(Pair(0, TextToSpeechState(IDLE, null)))
+        cleanupMediator.addSource(bindCount) { count ->
+            // these checks must be null safe to prevent the service from crashing
+            cleanupMediator.postValue(
+                    Pair(count ?: 0,
+                            cleanupMediator.value?.second ?: TextToSpeechState(IDLE, null)))
+        }
+        cleanupMediator.addSource(_state) { s ->
+            cleanupMediator.postValue(Pair(cleanupMediator.value?.first ?: 0,
+                    s ?: TextToSpeechState(IDLE, null)))
+        }
+        cleanupObserver = Observer { pair ->
+            if (pair != null && pair.first == 0 && pair.second.speakingState != SPEAKING
+                    && pair.second.speakingState != QUEUED) {
+                // If there are no bound clients, and speech isn't either currently happening, or queued to happen
+                // we should stop this service.
+                stopSelf()
+            }
+        }
+
+        cleanupMediator.observeForever(cleanupObserver)
+    }
+
+    override fun onCreate() {
+        LOGGER.info("onCreate called")
+        super.onCreate()
+        AndroidInjection.inject(this)
+    }
+
+    override fun onBind(intent: Intent?): IBinder {
+        val bindCountValue = (bindCount.value ?: 0) + 1
+        bindCount.postValue(bindCountValue)
+        LOGGER.debug("onBind() called bindCount = $bindCount")
+        return serviceBinder
+    }
+
+    override fun onUnbind(intent: Intent?): Boolean {
+        val bindCountValue = (bindCount.value ?: 1) - 1
+        bindCount.postValue(bindCountValue)
+        LOGGER.debug("onUnbind() called bindCount = $bindCount")
+        return super.onUnbind(intent)
+    }
+
+    override fun onDestroy() {
+        LOGGER.info("onDestroy called")
+        clear()
+        cleanupMediator.removeObserver(cleanupObserver)
+        super.onDestroy()
+    }
+
+    override fun onInit(status: Int) {
+        if (status == TextToSpeech.SUCCESS) {
+            val languageAvailable = textToSpeech?.isLanguageAvailable(Locale.getDefault())
+            // >= 0 means LANG_AVAILABLE, LANG_COUNTRY_AVAILABLE, or LANG_COUNTRY_VAR_AVAILABLE
+            if (languageAvailable != null && languageAvailable >= 0) {
+                textToSpeech?.let { tts ->
+                    tts.language = Locale.getDefault()
+                    tts.setOnUtteranceProgressListener(object : UtteranceProgressListener() {
+                        override fun onDone(text: String?) {
+                            LOGGER.info("TTS done speaking $text")
+                            val data = futureSpeechData
+                            if (data == null || !data.hasQueuedSpeech()) {
```

Comment:
This was another bug that was really odd. This block of code used to be wrapped in this conditional...

```
if (!tts.isSpeaking) {
    if (data == null || !data.hasQueuedSpeech()) { etc....
}
```

And VERY occasionally like 1 in 10, onDone was being called when tts.isSpeaking was returning true.  I'm guessing it was a threading edge case issue, but when I removed that unnecessary conditional, everything worked fine.